util/progress/progressui/printer.go

```diff
@@ -338,3 +353,52 @@ func limitString(s string, l int) string {
 	}
 	return s
 }
+
+// textPrinter prints the progress to a raw [io.Writer].
+type textPrinter interface {
+	// Print will print a message for the given build step at the index.
+	//
+	// If no index is associated with the message, a negative number will
+	// be used.
+	Print(w io.Writer, index int, msg string) error
+}
+
+// rawPrinter prints the text directly to the [io.Writer] with minor formatting.
+type rawPrinter struct{}
+
+func (*rawPrinter) Print(w io.Writer, index int, msg string) error {
+	if index < 0 {
+		_, err := fmt.Fprintf(w, "%s\n", msg)
+		return err
+	}
+	_, err := fmt.Fprintf(w, "#%d %s\n", index, msg)
+	return err
+}
+
+// jsonPrinter prints the text in JSON format.
+type jsonPrinter struct{}
+
+func (*jsonPrinter) Print(w io.Writer, index int, msg string) error {
+	if index < 0 && len(msg) == 0 {
+		// Skip empty messages for the json printer.
+		return nil
```

Comment:
I'm not sure if this is the correct decision since it might be making a choice for the user that we don't want to make.